### PR TITLE
Move block/state/warpc sync requests/responses to `ChainSync`

### PR DESCRIPTION
**Pull request by altonen**
This PR moves most of the code related to block requests away from `Protocol`/`Behaviour`/`NetworkWorker` to `ChainSync`. There is still a bit of block request code left relating to incoming peer connection/block announcements but those will be addressed in future PRs.

Resolves https://github.com/paritytech/substrate/issues/12133

polkadot companion: https://github.com/paritytech/polkadot/pull/6319
cumulus companion: https://github.com/paritytech/cumulus/pull/1899

### CODE REVIEW

**Review by bkchr — APPROVED** [review]
You added some functions as pub that don't need to be public? Otherwise I assume you just copied most of the code and didn't changed any logic?

**Comment by altonen** [comment]
The only change in logic is how the pending responses are stored. Previously they were stored together with the peer and the processing loop would go through all peers and process all ready responses. That was changed to push the pending responses to a `FuturesUnordered` and poll that instead.

File: client/network/sync/src/lib.rs [hunk 1]
```diff
@@ -2203,122 +1944,759 @@ where
...
+				log::warn!(
+					target: "sync",
+					"Trying to send warp sync request when no protocol is configured {:?}",
+					request,
+				);
+			},
+		}
+	}
+
+	pub fn on_block_response(
```
[the remark this answers is not in the document]
Comment (altonen) [reply client/network/sync/src/lib.rs:2134]: It doens't

File: client/network/sync/Cargo.toml [hunk 2]
```diff
@@ -18,6 +18,7 @@ prost-build = "0.11"
 
 [dependencies]
 array-bytes = "4.1"
+async-trait = "0.1"
```
[the remark this answers is not in the document]
Comment (altonen) [reply client/network/sync/Cargo.toml:21]: In particular reason why not just let cargo select the patch version?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2067122

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2067117

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2067116

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2067110

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2067109

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-subkey
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2067111

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2067154

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2067193

**Review by arkpar — APPROVED** [review]

File: client/network/sync/Cargo.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (bkchr) [reply client/network/sync/Cargo.toml]: Let's assume we are using the patch version of a crate that introduces a new function, but some downstream user doesn't use the same version. They could get a compilation error, which would be easy to solve, but still an error. If we directly force this version, downstream will be "forced" to use this patch version as well.

**Pull request merged**